control/grpc.py

```diff
@@ -2530,6 +2560,156 @@ def remove_host_safe(self, request, context):
     def remove_host(self, request, context=None):
         return self.execute_grpc_function(self.remove_host_safe, request, context)
 
+    def change_host_keys_safe(self, request, context):
+        """Changes host's inband authentication keys."""
+
+        peer_msg = self.get_peer_message(context)
+        failure_prefix=f"Failure changing keys for host {request.host_nqn} on subsystem {request.subsystem_nqn}"
+        self.logger.info(
+            f"Received request to change inband authentication keys for host {request.host_nqn} on subsystem {request.subsystem_nqn}, dhchap: {request.dhchap_key}, dhchap controller: {request.dhchap_ctrlr_key}, context: {context}{peer_msg}")
+
+        if request.host_nqn == "*":
+            errmsg=f"{failure_prefix}: Host NQN can't be '*'"
+            self.logger.error(f"{errmsg}")
+            return pb2.req_status(status=errno.EINVAL, error_message=errmsg)
+
+        if not GatewayState.is_key_element_valid(request.host_nqn):
+            errmsg = f"{failure_prefix}: Invalid host NQN \"{request.host_nqn}\", contains invalid characters"
+            self.logger.error(f"{errmsg}")
+            return pb2.req_status(status = errno.EINVAL, error_message = errmsg)
+
+        if not GatewayState.is_key_element_valid(request.subsystem_nqn):
+            errmsg = f"{failure_prefix}: Invalid subsystem NQN \"{request.subsystem_nqn}\", contains invalid characters"
+            self.logger.error(f"{errmsg}")
+            return pb2.req_status(status = errno.EINVAL, error_message = errmsg)
+
+        if self.verify_nqns:
+            rc = GatewayUtils.is_valid_nqn(request.host_nqn)
+            if rc[0] != 0:
+                errmsg = f"{failure_prefix}: {rc[1]}"
+                self.logger.error(f"{errmsg}")
+                return pb2.req_status(status = rc[0], error_message = errmsg)
+
+        if GatewayUtils.is_discovery_nqn(request.host_nqn):
+            errmsg=f"{failure_prefix}: Can't use a discovery NQN as host's"
+            self.logger.error(f"{errmsg}")
+            return pb2.req_status(status=errno.EINVAL, error_message=errmsg)
+
+        if request.dhchap_ctrlr_key and not request.dhchap_key:
+            errmsg=f"{failure_prefix}: DH-HMAC-CHAP controller key can only be used with a DH-HMAC-CHAP key"
+            self.logger.error(f"{errmsg}")
+            return pb2.req_status(status=errno.EINVAL, error_message=errmsg)
+
+        host_already_exist = self.matching_host_exists(context, request.subsystem_nqn, request.host_nqn)
+        if not host_already_exist and context:
```

Comment:
not sure about the `and context` part— is it OK if the host doesn’t exist for an OMAP update?

Reply:
The matching_host_exists() function only works when we are not on update. It is not that it's OK the host doesn't exist, but in case of an update we should have already verified the host was there.

Reply:
I didn't want to make too many changes to existing code for this as it's that important. If the host is not there we'll get an error from SPDK error. The only purpose of this test is to have a nicer looking error message.